File: csharp/data_structure/stack/Stack.cs

using System;
using System.Collections.Generic;

namespace DataStructure.Stack;

public class Stack<T>
{
    private readonly List<T> _items;

    public Stack()
    {
        _items = new List<T>();
    }

    public void Push(T item) => _items.Add(item);
    public T Pop()
    {
        if (IsEmpty())
            throw new InvalidOperationException("pop from empty stack");

        int lastIndex = _items.Count - 1;
        T value = _items[lastIndex];
        _items.RemoveAt(lastIndex);
        return value;
    }

    public T Peek()
    {
        if (IsEmpty())
            throw new InvalidOperationException("peek from empty stack");

        return _items[^1]; // same as _items[_items.Count - 1]
    }

    public bool IsEmpty() => _items.Count == 0;

    public int Size() => _items.Count;

    public override string ToString() => $"Stack({_items.Count} elements)";
}

File: java/data_structures/stack/Stack.java
package data_structures.stack;

import java.util.ArrayList;
import java.util.NoSuchElementException;

public class Stack<E> {

    private final ArrayList<E> items;

    public Stack() {
        this.items = new ArrayList<>();
    }

    public void push(E item) {
        items.add(item);
    }

    public E pop() {
        if (isEmpty()) {
            throw new NoSuchElementException("pop from empty stack");
        }
        return items.remove(items.size() - 1);
    }

    public E peek() {
        if (isEmpty()) {
            throw new NoSuchElementException("peek from empty stack");
        }
        return items.get(items.size() - 1);
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    public int size() {
        return items.size();
    }

    @Override
    public String toString() {
        return "Stack(" + items.toString() + ")";
    }
}
